src/libraries/FluidLib.sol

```diff
@@ -0,0 +1,196 @@
+                RateLimitHelpers.makeAssetDestinationKey(
+                    params.depositSLRateLimitId,
+                    token0,
+                    params.smartLending
+                ),
+                params.token0Amount
+            );
+
+            // Approve asset to token from the proxy (assumes the proxy has enough of the asset).
+            _approve(
+                params.proxy,
+                params.smartLending,
+                token0,
+                params.token0Amount
+            );
+        }
+        if (params.token1Amount > 0) {
+            address token1 = smartLending.TOKEN1();
+
+            params.rateLimits.triggerRateLimitDecrease(
+                RateLimitHelpers.makeAssetDestinationKey(
+                    params.depositSLRateLimitId,
+                    token1,
+                    params.smartLending
+                ),
+                params.token1Amount
+            );
+
+            // Approve asset to token from the proxy (assumes the proxy has enough of the asset).
+            _approve(
+                params.proxy,
+                params.smartLending,
+                token1,
+                params.token1Amount
+            );
+        }
+
+        (, shares) = abi.decode(
+            params.proxy.doCall(
+                params.smartLending,
+                abi.encodeCall(
+                    smartLending.deposit,
+                    (
+                        params.token0Amount,
+                        params.token1Amount,
+                        params.minShares,
+                        address(params.proxy)
+                    )
+                )
+            ),
+            (uint256, uint256)
+        );
+    }
+
+    function withdrawSmartLending(
+        WithdrawSmartLendingParams calldata params
+    ) external returns (uint256 shares) {
+        IFluidSmartLending smartLending = IFluidSmartLending(
+            params.smartLending
+        );
+
+        bytes32 rateLimitKey = RateLimitHelpers.makeAssetKey(
+            params.withdrawSLRateLimitId,
+            params.smartLending
+        );
+
+        params.rateLimits.triggerRateLimitDecrease(
+            rateLimitKey,
+            params.maxShares
+        );
+
+        // Withdraw asset from a token, decode resulting shares.
+        // Assumes proxy has adequate token shares.
+        (, shares) = abi.decode(
+            params.proxy.doCall(
+                params.smartLending,
+                abi.encodeCall(
+                    smartLending.withdraw,
+                    (
+                        params.token0Amount,
+                        params.token1Amount,
+                        params.maxShares,
+                        address(params.proxy)
+                    )
+                )
+            ),
+            (uint256, uint256)
+        );
+
+        uint256 sharesDiff = params.maxShares - shares; // actual withdrawn shares is always <= maxShares
+        if (sharesDiff > 0) {
```

Comment:
an alternative approach here to increasing could be, decrease only by e.g. 99% above (assuming max ever slippage configurable possible is 1%) and then decrease further here to the actual amount. The current approach is slightly more restrictive though.